src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintHelper.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using Microsoft.AspNet.OData.Common;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    internal static class DerivedTypeConstraintHelper
+    {
+        internal static void ValidateAndAddConstraint(this ISet<Type> Constraints, Type ClrType, Type typeToAdd)
```

Comment:
>ValidateAndAddConstraint [](start = 29, length = 24)

Do you really want to add this to any ISet?  Wouldn't it be better to create a DerivedTypeConstraintSet<T> : HashSet<T> that implements this method, and then use that to hold derived type constraints?